File: src/pixel_art_mcp/imaging/pet.py
"""Export a pixel-index custom-pet package: manifest.json + a 96x96 pet.png.

See docs/custom-asset-zip-contract.md (pixel-agents-hq/index): row 0 (y=0..32) is
walkDown[0..2]+idleDown[0..2] at 16x32 each; row 1 (y=32..64) is the same for `up`;
row 2 (y=64..96) is walkRight[0..2] at 32x32 each -- rendered at double width by
blender/runner.py's per-angle resolution_x (see angle_widths() below, shared with
jobs/worker.py). idleRight doesn't exist in the contract: idle-state right-facing
frames are still rendered (one shared render pass covers every state at every angle)
but are simply not used here.
"""

import json
from pathlib import Path
from typing import Any
from zipfile import ZIP_DEFLATED, ZipFile

from PIL import Image

from pixel_art_mcp.imaging.pixels import pixelate
from pixel_art_mcp.models import DomainError, RenderOptions

PET_PNG_SIZE = (96, 96)
PET_FRAME_HEIGHT = 32
PET_NARROW_WIDTH = 16
PET_WIDE_WIDTH = 32
PET_WIDE_ANGLE = 90.0
# Documented (not pixel-index-enforced) per-pet-PNG compatibility cap -- see the
# contract doc's note on upstream's own MAX_PET_PNG_SIZE.
MAX_PET_PNG_BYTES = 512 * 1024


def angle_widths(width: int, angles: list[float]) -> list[int]:
    """Per-angle render width for a pet job: double width for the `right` (90deg) row,
    matching the contract's wider side-view canvas. Used both by worker.py (to tell
    the Blender renderer, from the raw job options dict) and by this module (to decode
    the resulting raw PNGs, from a validated RenderOptions)."""
    return [PET_WIDE_WIDTH if angle == PET_WIDE_ANGLE else width for angle in angles]


def export_pet_sheet(
    raw_dir: Path,
    output_dir: Path,
    manifest: dict[str, Any],
    options: RenderOptions,
    project_id: str,
    revision_id: str,
) -> None:
    target = options.pet
    states = options.states
    assert target is not None and states is not None
    walk = next(state for state in states if state.id == "walk")
    idle = next(state for state in states if state.id == "idle")

    entries = manifest["frames"]
    rendered_frames = options.render_frames()
    if len(entries) != len(options.angles) * len(rendered_frames):
        raise DomainError("Blender returned an incomplete pet frame sequence")
    by_key = {(entry["angle"], entry["frame"]): entry for entry in entries}
    widths = dict(zip(options.angles, angle_widths(options.width, options.angles), strict=True))

    def load(angle: float, frame: int, width: int) -> Image.Image:
        entry = by_key.get((angle, frame))
        if entry is None:
            raise DomainError("Blender did not render a required pet frame")
        path = (raw_dir / entry["filename"]).resolve()
        if not path.is_relative_to(raw_dir.resolve()) or path.suffix != ".png":
            raise DomainError("Invalid render output path")
        with Image.open(path) as source:
            expected_size = (width * options.supersampling, options.height * options.supersampling)
            if source.size != expected_size:
                raise DomainError("Blender returned unexpected image dimensions")
            return source.convert("RGBA")

    # The ordered list of (angle, frame, width) cells this pet needs. One shared
    # palette is derived across all of them (via pixelate()) so the down/up rows and
    # the wider right row share consistent colors.
    cells: list[tuple[float, int, int]] = []
    for angle in (0.0, 180.0):
        for frame in walk.frames() + idle.frames():
            cells.append((angle, frame, widths[angle]))
    for frame in walk.frames():
        cells.append((PET_WIDE_ANGLE, frame, widths[PET_WIDE_ANGLE]))

    sources = (load(angle, frame, width) for angle, frame, width in cells)
    sizes = [(width, options.height) for _, _, width in cells]
    rendered, palette = pixelate(sources, options, sizes=sizes)
    by_cell = {
        (angle, frame): image for (angle, frame, _), image in zip(cells, rendered, strict=True)
    }

    sheet = Image.new("RGBA", PET_PNG_SIZE)

    def paste_row(y: int, angle: float, frames: list[int], width: int) -> None:
        x = 0
        for frame in frames:
            sheet.paste(by_cell[(angle, frame)], (x, y))
            x += width

    paste_row(0, 0.0, walk.frames() + idle.frames(), PET_NARROW_WIDTH)
    paste_row(PET_FRAME_HEIGHT, 180.0, walk.frames() + idle.frames(), PET_NARROW_WIDTH)
    paste_row(PET_FRAME_HEIGHT * 2, PET_WIDE_ANGLE, walk.frames(), PET_WIDE_WIDTH)

    output_dir.mkdir(parents=True, exist_ok=True)
    package_root = output_dir / "pixel-agents-pet"
    directory = package_root / target.asset_id
    directory.mkdir(parents=True)
    (directory / "manifest.json").write_text(
        json.dumps({"id": target.asset_id, "name": target.name}, indent=2), encoding="utf-8"
    )
    png_path = directory / "pet.png"
    sheet.save(png_path)
    if png_path.stat().st_size > MAX_PET_PNG_BYTES:
        raise DomainError(
            f"Generated pet.png exceeds pixel-agents' {MAX_PET_PNG_BYTES}-byte per-pet "
            "compatibility cap; reduce colors or simplify the model"
        )

    archive_name = "pixel-agents-pet.zip"
    with ZipFile(output_dir / archive_name, "w", ZIP_DEFLATED) as archive:
        for path in sorted(directory.iterdir()):
            archive.write(path, path.relative_to(package_root))

    metadata = {
        "schema_version": 1,
        "project_id": project_id,
        "revision_id": revision_id,
        "image": f"pixel-agents-pet/{target.asset_id}/pet.png",
        "size": list(PET_PNG_SIZE),
        "palette": palette,
        "transparent": True,
        "settings": options.model_dump(),
        "pet": {
            "asset_id": target.asset_id,
            "name": target.name,
            "manifest": f"pixel-agents-pet/{target.asset_id}/manifest.json",
            "archive": archive_name,
            "width": PET_PNG_SIZE[0],
            "height": PET_PNG_SIZE[1],
            "walk_frames": walk.frames(),
            "idle_frames": idle.frames(),
            "install": "Upload pet.png + manifest.json to pixel-index with assetKind=pet.",
        },
        "camera": manifest["camera"],
        "blender_version": manifest["blender_version"],
    }
    (output_dir / "spritesheet.json").write_text(json.dumps(metadata, indent=2), encoding="utf-8")
    with ZipFile(output_dir / "sprites.zip", "w", ZIP_DEFLATED) as archive:
        for path in sorted(output_dir.rglob("*")):
            if path.is_file() and path.name != "sprites.zip":
                archive.write(path, path.relative_to(output_dir))

File: packages/imaging/src/pet.ts
/**
 * Export a pixel-index custom-pet package: manifest.json + a 96x96 pet.png. Verbatim port of
 * `src/pixel_art_mcp/imaging/pet.py` (149 lines).
 *
 * See docs/custom-asset-zip-contract.md (pixel-agents-hq/index): row 0 (y=0..32) is
 * walkDown[0..2]+idleDown[0..2] at 16x32 each; row 1 (y=32..64) is the same for `up`; row 2
 * (y=64..96) is walkRight[0..2] at 32x32 each -- rendered at double width by the per-angle
 * render width (see `angleWidths` below, shared with the job worker that decides render
 * resolution). idleRight doesn't exist in the contract: idle-state right-facing frames are still
 * rendered (one shared render pass covers every state at every angle) but simply not used here.
 */

import fs from "node:fs";
import path from "node:path";

import { DomainError, renderOptionsRenderFrames, renderStateFrames, type RenderOptions } from "@pixel-art-mcp/schema";

import { createImage, isWithinDirectory, pasteFull, readPng, writePng, type RGBAImage } from "./image.js";
import { at, defined } from "./internal.js";
import { writeSpritesZip, zipDirectory } from "./pack-zip.js";
import { pixelate } from "./pixels.js";

export const PET_PNG_SIZE: readonly [number, number] = [96, 96];
export const PET_FRAME_HEIGHT = 32;
export const PET_NARROW_WIDTH = 16;
export const PET_WIDE_WIDTH = 32;
export const PET_WIDE_ANGLE = 90;
/** Documented (not pixel-index-enforced) per-pet-PNG compatibility cap -- see the contract doc's
 * note on upstream's own MAX_PET_PNG_SIZE. */
export const MAX_PET_PNG_BYTES = 512 * 1024;

/** Per-angle render width for a pet job: double width for the `right` (90deg) row, matching the
 * contract's wider side-view canvas. Used both by the job worker (to tell the renderer, from the
 * raw job options) and by this module (to decode the resulting raw PNGs, from a validated
 * RenderOptions). */
export function angleWidths(width: number, angles: readonly number[]): number[] {
  return angles.map((angle) => (angle === PET_WIDE_ANGLE ? PET_WIDE_WIDTH : width));
}

interface PetManifestFrame {
  angle: number;
  frame: number;
  filename: string;
}

interface PetManifest {
  frames: readonly PetManifestFrame[];
  camera: Record<string, unknown>;
}

function cellKey(angle: number, frame: number): string {
  return `${String(angle)}:${String(frame)}`;
}

export function exportPetSheet(
  rawDir: string,
  outputDir: string,
  manifest: PetManifest,
  options: RenderOptions,
  projectId: string,
  revisionId: string,
): void {
  const target = options.pet;
  const states = options.states;
  if (!target || !states) {
    throw new Error("exportPetSheet requires options.pet and options.states");
  }
  const walk = defined(
    states.find((state) => state.id === "walk"),
    "pet walk state",
  );
  const idle = defined(
    states.find((state) => state.id === "idle"),
    "pet idle state",
  );

  const entries = manifest.frames;
  const renderedFrames = renderOptionsRenderFrames(options);
  if (entries.length !== options.angles.length * renderedFrames.length) {
    throw new DomainError("Render output has an incomplete pet frame sequence");
  }
  const byKey = new Map<string, PetManifestFrame>();
  for (const entry of entries) byKey.set(cellKey(entry.angle, entry.frame), entry);
  const widthsByAngle = new Map<number, number>();
  angleWidths(options.width, options.angles).forEach((width, index) => {
    widthsByAngle.set(at(options.angles, index), width);
  });

  const resolvedRawDir = path.resolve(rawDir);
  function load(angle: number, frame: number, width: number): RGBAImage {
    const entry = byKey.get(cellKey(angle, frame));
    if (!entry) throw new DomainError("Render output is missing a required pet frame");
    const resolvedPath = path.resolve(rawDir, entry.filename);
    if (!isWithinDirectory(resolvedRawDir, resolvedPath) || path.extname(resolvedPath) !== ".png") {
      throw new DomainError("Invalid render output path");
    }
    const source = readPng(resolvedPath);
    const expectedWidth = width * options.supersampling;
    const expectedHeight = options.height * options.supersampling;
    if (source.width !== expectedWidth || source.height !== expectedHeight) {
      throw new DomainError("Render output has unexpected image dimensions");
    }
    return source;
  }

  const walkFrames = renderStateFrames(walk);
  const idleFrames = renderStateFrames(idle);

  // The ordered list of (angle, frame, width) cells this pet needs. One shared palette is
  // derived across all of them (via `pixelate()`) so the down/up rows and the wider right row
  // share consistent colors.
  const cells: [number, number, number][] = [];
  for (const angle of [0, 180]) {
    for (const frame of [...walkFrames, ...idleFrames]) {
      cells.push([angle, frame, defined(widthsByAngle.get(angle), `render width for angle ${String(angle)}`)]);
    }
  }
  for (const frame of walkFrames) {
    cells.push([
      PET_WIDE_ANGLE,
      frame,
      defined(widthsByAngle.get(PET_WIDE_ANGLE), "render width for the wide angle"),
    ]);
  }

  const sources = cells.map(([angle, frame, width]) => load(angle, frame, width));
  const sizes: [number, number][] = cells.map(([, , width]) => [width, options.height]);
  const [rendered, palette] = pixelate(sources, options, sizes);
  const byCell = new Map<string, RGBAImage>();
  cells.forEach(([angle, frame], index) => byCell.set(cellKey(angle, frame), at(rendered, index)));

  const sheet = createImage(PET_PNG_SIZE[0], PET_PNG_SIZE[1]);

  function pasteRow(y: number, angle: number, frames: readonly number[], width: number): void {
    let x = 0;
    for (const frame of frames) {
      pasteFull(sheet, defined(byCell.get(cellKey(angle, frame)), "pet cell"), x, y);
      x += width;
    }
  }
  pasteRow(0, 0, [...walkFrames, ...idleFrames], PET_NARROW_WIDTH);
  pasteRow(PET_FRAME_HEIGHT, 180, [...walkFrames, ...idleFrames], PET_NARROW_WIDTH);
  pasteRow(PET_FRAME_HEIGHT * 2, PET_WIDE_ANGLE, walkFrames, PET_WIDE_WIDTH);

  fs.mkdirSync(outputDir, { recursive: true });
  const packageRoot = path.join(outputDir, "pixel-agents-pet");
  const directory = path.join(packageRoot, target.asset_id);
  fs.mkdirSync(directory, { recursive: true });
  fs.writeFileSync(
    path.join(directory, "manifest.json"),
    JSON.stringify({ id: target.asset_id, name: target.name }, null, 2),
    "utf-8",
  );
  const pngPath = path.join(directory, "pet.png");
  writePng(pngPath, sheet);
  if (fs.statSync(pngPath).size > MAX_PET_PNG_BYTES) {
    throw new DomainError(
      `Generated pet.png exceeds pixel-agents' ${String(MAX_PET_PNG_BYTES)}-byte per-pet ` +
        "compatibility cap; reduce colors or simplify the model",
    );
  }

  const archiveName = "pixel-agents-pet.zip";
  zipDirectory(packageRoot, path.join(outputDir, archiveName));

  const metadata = {
    schema_version: 1,
    project_id: projectId,
    revision_id: revisionId,
    image: `pixel-agents-pet/${target.asset_id}/pet.png`,
    size: [PET_PNG_SIZE[0], PET_PNG_SIZE[1]],
    palette,
    transparent: true,
    settings: options,
    pet: {
      asset_id: target.asset_id,
      name: target.name,
      manifest: `pixel-agents-pet/${target.asset_id}/manifest.json`,
      archive: archiveName,
      width: PET_PNG_SIZE[0],
      height: PET_PNG_SIZE[1],
      walk_frames: walkFrames,
      idle_frames: idleFrames,
      install: "Upload pet.png + manifest.json to pixel-index with assetKind=pet.",
    },
    camera: manifest.camera,
  };
  fs.writeFileSync(path.join(outputDir, "spritesheet.json"), JSON.stringify(metadata, null, 2), "utf-8");
  writeSpritesZip(outputDir);
}
